test/e2e/framework/framework.go

```diff
@@ -48,69 +48,91 @@ var _ ClusterInterface = &Framework{}
-		f.e2eArtifactsDir = path.Join(TestContext.ArtifactsDir, "e2e")
-		err = os.Mkdir(f.e2eArtifactsDir, 0777)
+		e2eArtifactsDir = path.Join(TestContext.ArtifactsDir, "e2e")
+		err = os.Mkdir(e2eArtifactsDir, 0777)
 		if !os.IsExist(err) {
 			o.Expect(err).NotTo(o.HaveOccurred())
 		}
 	}
 
-	o.Expect(TestContext.RestConfigs).NotTo(o.BeEmpty())
-	for i, restConfig := range TestContext.RestConfigs {
-		clusterName := fmt.Sprintf("%s-%d", f.name, i)
+	f := &Framework{
+		FullClient: FullClient{},
+
+		name: names.SimpleNameGenerator.GenerateName(fmt.Sprintf("%s-", namePrefix)),
+
+		workerClusters: map[string]*Cluster{},
+	}
 
-		clusterE2EArtifactsDir := ""
-		if len(f.e2eArtifactsDir) != 0 {
-			clusterE2EArtifactsDir = path.Join(f.e2eArtifactsDir, fmt.Sprintf("cluster-%d", i))
-			err = os.Mkdir(clusterE2EArtifactsDir, 0777)
+	clusterE2EArtifactsDir := ""
+	if len(e2eArtifactsDir) != 0 {
+		clusterE2EArtifactsDir = path.Join(e2eArtifactsDir, "cluster")
+		err = os.Mkdir(clusterE2EArtifactsDir, 0777)
+		if !os.IsExist(err) {
+			o.Expect(err).NotTo(o.HaveOccurred())
+		}
+	}
+	f.cluster = newCluster(f.name, clusterE2EArtifactsDir, TestContext.RestConfig, f.CommonLabels())
+	f.FullClient.AdminClient.Config = f.cluster.AdminClientConfig()
+
+	for workerClusterKey, workerClusterRestConfig := range TestContext.WorkerRestConfigs {
+		workerClusterE2EArtifactsDir := ""
+		if len(e2eArtifactsDir) != 0 {
+			workerClusterE2EArtifactsDir = path.Join(e2eArtifactsDir, "workers", workerClusterKey)
+			err = os.MkdirAll(workerClusterE2EArtifactsDir, 0777)
 			if !os.IsExist(err) {
 				o.Expect(err).NotTo(o.HaveOccurred())
 			}
 		}
-
-		c := NewCluster(
-			clusterName,
-			clusterE2EArtifactsDir,
-			restConfig,
-			func(ctx context.Context, adminClient kubernetes.Interface, adminClientConfig *restclient.Config) (*corev1.Namespace, Client) {
-				return CreateUserNamespace(ctx, clusterName, f.CommonLabels(), adminClient, adminClientConfig)
-			},
-		)
-		f.clusters = append(f.clusters, c)
+		workerClusterName := fmt.Sprintf("%s-%s", f.name, workerClusterKey)
+		f.workerClusters[workerClusterKey] = newCluster(workerClusterName, workerClusterE2EArtifactsDir, workerClusterRestConfig, f.CommonLabels())
 	}
 
-	f.FullClient.AdminClient.Config = f.defaultCluster().AdminClientConfig()
-
 	g.BeforeEach(f.beforeEach)
 	g.JustAfterEach(f.justAfterEach)
 	g.AfterEach(f.afterEach)
 
 	return f
 }
 
-func (f *Framework) Cluster(idx int) ClusterInterface {
-	o.Expect(idx).NotTo(o.BeNumerically(">=", len(f.clusters)))
-	return f.clusters[idx]
+func newCluster(name, artifactsDir string, restConfig *restclient.Config, namespaceLabels map[string]string) *Cluster {
+	return NewCluster(
+		name,
+		artifactsDir,
+		restConfig,
+		func(ctx context.Context, adminClient kubernetes.Interface, adminClientConfig *restclient.Config) (*corev1.Namespace, Client) {
+			return CreateUserNamespace(ctx, name, namespaceLabels, adminClient, adminClientConfig)
+		},
+	)
+}
```

Comment:
Why not modify `NewCluster`'s signature to accept the same set of arguments? It's not obvious to me why two separate constructors are needed, where a private one calls a public one.

Reply:
done